File transfer

agneborn98 opened the pull request:
This PR transfers the repo from my personal one to the team one. Everything should work here as long as the instructions are followed.

agneborn98 committed:
Initial transfer from personal repo

agneborn98 committed:
Update README.md

agneborn98 committed:
Minor updates

agneborn98 committed:
Update sample.conf

agneborn98 committed:
Update README.md

agneborn98 committed:
Update README.md

agneborn98 committed:
Update README.md

agneborn98 committed:
Update README.md

agneborn98 committed:
Updated README and telegraf.conf

I added an explanation on how to configure the telegraf.conf file, and added it to .gitignore.

agneborn98 committed:
Delete telegraf.conf

Removed telegraf.conf as requested. The file is also added to .gitignore.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
-External Telegraf plugin for writing time series to Apache Arrow Flight
+# Apache Arrow Flight External Output Plugin
+
+This plugin writes to [ModelarDB-RS](https://github.com/ModelarData/ModelarDB-RS) via the [Apache Arrow Flight protocol](https://arrow.apache.org/docs/format/Flight.html).
```
I think "ModelarDB" should be used instead of "ModelarDB-RS" to be consistent with the other places we references the repository.

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
-External Telegraf plugin for writing time series to Apache Arrow Flight
+# Apache Arrow Flight External Output Plugin
+
+This plugin writes to [ModelarDB-RS](https://github.com/ModelarData/ModelarDB-RS) via the [Apache Arrow Flight protocol](https://arrow.apache.org/docs/format/Flight.html).
```
Instead of stating that it is for "ModelarDB" I suggest we state it is as a general purpose output plugin for "Apache Arrow Flight" that only implements support for the simplest schema supported by legacy JVM and current Rust versions of ModelarDB at the moment (maybe in known issues). We have to implement support for arbitrary schemas anyway to support ingesting data points for multivariate time series that contain a single times stamp, a variable number of fields, and a variable number of tags, anyway. I also think it will be easier to encourage other to contribute to it and hopefully upstream it later if we do not make hardcode it to only work with ModelarDB. What do you think @CGodiksen and @chrthomsen?

agneborn98 replied:
I tend to agree. It just seemed very premature to already call it a "general purpose" because it was so specific to our use case. I have made the suggested changes, but I don't see us getting any contributions until the plugin has been made totally general purpose and made internal via a [pull request](https://github.com/influxdata/telegraf/blob/master/CONTRIBUTING.md#contributing-code).

CGodiksen replied:
I agree that we should aim the documentation at being general purpose and might as well do this now. As long as we don't make any contributions to Telegraf before it is completely done.

skejserjensen replied:
With the exception of a few minor things for which I have added comments, the updated README is fine for me. @CGodiksen please resolve this discussion if agree.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
+
+## Arrow Schema
+
+The plugin currently supports inserting data points with the following schema:
```
Maybe add that support for arbitrary schemas are planned since we need it anyway for multivariate time series.

agneborn98 replied:
Added.

skejserjensen commented on README.md:
````diff
@@ -1,2 +1,56 @@
+
+The plugin currently supports inserting data points with the following schema:
+
+```rust
````
Maybe write the schema using a simple textual format, e.g., the one printed by `fmt`, instead of code. If code is required maybe use Go code instead of Rust code.

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
+2. Build the binary:
+    * Windows: `go build -o binary/flight.exe cmd/main.go`
+    * Linux/macOS: `go build -o binary/flight cmd/main.go`
+3. Download the latest version of [telegraf](https://github.com/influxdata/telegraf/releases) for your platform. (To see which platform is needed, run the following command: `go env GOOS GOARCH`)
```
Maybe link to the [InfluxData](https://portal.influxdata.com/downloads/) website as it also provides installation instruction, although, some of the binaries are only available from GitHub.

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
+    * Windows: `go build -o binary/flight.exe cmd/main.go`
+    * Linux/macOS: `go build -o binary/flight cmd/main.go`
+3. Download the latest version of [telegraf](https://github.com/influxdata/telegraf/releases) for your platform. (To see which platform is needed, run the following command: `go env GOOS GOARCH`)
+4. Extract the telegraf executable and configuration to the repository folder.
```
Maybe remove if the link above is changed to InfluxData.

agneborn98 replied:
I'm not sure I understand, why does a new link change the instructions? 

Should I link to [this](https://docs.influxdata.com/telegraf/v1.21/external_plugins/shim/)? I think that seems like a bit too many instructions to just build and run the plugin.

skejserjensen replied:
My point was that bullet 4 may not be necessary if we link to https://portal.influxdata.com/downloads/ since platform and format specific installation instructions are provided on that website. However, if you think the instructions are more clear if we keep bullet 4. then that is also Ok for me.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
+3. Download the latest version of [telegraf](https://github.com/influxdata/telegraf/releases) for your platform. (To see which platform is needed, run the following command: `go env GOOS GOARCH`)
+4. Extract the telegraf executable and configuration to the repository folder.
+5. Configure the telegraf.conf file:
+   * Search for `execd`, remove the comment in front of the tag `[[outputs.execd]]` and the option `command`.
```
Maybe replace "Search of `execd`" with "In `telegraf.conf`".

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
+4. Extract the telegraf executable and configuration to the repository folder.
+5. Configure the telegraf.conf file:
+   * Search for `execd`, remove the comment in front of the tag `[[outputs.execd]]` and the option `command`.
+     * Windows: Assign the following to `command`: `["./binary/flight.exe", "-config", "./plugins/output/flight/sample.conf"]` 
```
Maybe provide the full `command` line so it is clear what assign means, e.g., with "... to `command` so the resulting lines becomes: ....". Also maybe replace specific paths with text indicating what the path should be, e.g, replace "./binary/flight.exe" with "/path/to/flight.exe".

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
+5. Configure the telegraf.conf file:
+   * Search for `execd`, remove the comment in front of the tag `[[outputs.execd]]` and the option `command`.
+     * Windows: Assign the following to `command`: `["./binary/flight.exe", "-config", "./plugins/output/flight/sample.conf"]` 
+     * Linux/macOS: Assign the following to `command`: `["./binary/flight", "-config", "./plugins/output/flight/sample.conf"]` 
```
Maybe provide the full `command` line so it is clear what assign means, e.g., with "... to `command` so the resulting lines becomes: ....". Maybe replace specific paths with text indicating what the path should be, e.g, replace "./binary/flight" with "/path/to/flight".

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
+     * Windows: Assign the following to `command`: `["./binary/flight.exe", "-config", "./plugins/output/flight/sample.conf"]` 
+     * Linux/macOS: Assign the following to `command`: `["./binary/flight", "-config", "./plugins/output/flight/sample.conf"]` 
+   * Configure any input plugin to consume metrics. (the metric must adhere to the schema presented in this README)
+6. Run the plugin using telegraf: `telegraf --config telegraf.conf --input-filter mqtt_consumer --output-filter execd`
```
Must `mqtt_consumer` always be included even if MQTT is noted used? If not maybe specify that input filter should be whatever the user have configured Telegraf to use.

agneborn98 replied:
No, you are right. As you can probably see, the readme wasn't written as a general purpose plugin initially, but with a very specific use case in mind.

I have changed this to better reflect a general-purpose plugin.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
+
+
+## Configuration
+
```
Maybe add some text describing what this configuration file is (e.g., is it a part of `telegraf.conf` or something separate), how to fill it out, and how to use it with the plugin.

agneborn98 replied:
I have added a short description with a link to the original path of the sample.conf.

I think a whole guide on how to fill it out seems a bit excessive, the small descriptions of each field should surmise for the skill level of the end users of the plugin. 

How to use it with the plugin is already described in the "To build the binary and run the plugin:"
If you think further explanation is required, please elaborate on this a bit more.

skejserjensen commented on README.md:
````diff
@@ -1,2 +1,56 @@
+```toml @sample.conf
+# Configuration for Arrow Flight to send metrics to.
+[[outputs.flight]]
+    ## Location to connect to.
````
If '#' and '##' have the same effect maybe use '#' everywhere for consistency.

agneborn98 replied:
As stated in the [sample config style guide](https://github.com/influxdata/telegraf/blob/master/docs/developers/SAMPLE_CONFIG.md#style), documentation should be double commented so I have changed it accordingly.

skejserjensen commented on README.md:
````diff
@@ -1,2 +1,56 @@
+```
+## Known issues
+
+* The descriptor is attached every time a payload is written to `DoPut()`.
````
Maybe specify why that is an issue, e.g., by linking to the part of the Apache Arrow Flight documentation that states it should not happen.

agneborn98 replied:
This is no longer an issue and has been removed from the known issues.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
+## Known issues
+
+* The descriptor is attached every time a payload is written to `DoPut()`.
+* `GetSchema()` relies on a "hack" in ModelarDB-RS, because a bug is present in the Rust implementation of Apache Arrow Flight where the schema is not serialized properly.
```
Maybe state that it is not compatible with the Rust implementation of Apache Arrow Flight implementation as it needs `GetSchema()` and the Rust implementation serializes the schema differently then everybody else and link to the issue.

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
````diff
@@ -1,2 +1,56 @@
+    ## Table in which to insert time series data.
+    table = "data"
+```
+## Known issues
````
Maybe "issues and limitations" since not everything listed here are issues.

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
+
+* The descriptor is attached every time a payload is written to `DoPut()`.
+* `GetSchema()` relies on a "hack" in ModelarDB-RS, because a bug is present in the Rust implementation of Apache Arrow Flight where the schema is not serialized properly.
+* The schema is retrieved from the flight server, but there is currently nothing implemented to handle a schema other than the one described above, because a builder has to manually be initialized for the type of each field.
```
Is the amount of detailed necessary? Maybe just specify that the schema listed above is currently the only one supported.

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,56 @@
+* The descriptor is attached every time a payload is written to `DoPut()`.
+* `GetSchema()` relies on a "hack" in ModelarDB-RS, because a bug is present in the Rust implementation of Apache Arrow Flight where the schema is not serialized properly.
+* The schema is retrieved from the flight server, but there is currently nothing implemented to handle a schema other than the one described above, because a builder has to manually be initialized for the type of each field.
+* Testing is still unstable.
```
It is not clear what this means, is the testing not comprehensive, are some of the tests flaky, etc.

agneborn98 replied:
The testing was half-baked, but it runs now if the instructions are followed.

skejserjensen commented on cmd/main.go:
```diff
@@ -0,0 +1,59 @@
+	"os"
+	"time"
+
+	// TODO: import your plugins
```
Are the comments in this file still relevant, they seem to be from a template. If they are no longer relevant they should probably be removed.

agneborn98 replied:
Are some of the comments in the file not relevant? Sure, I believe I have removed those now.

I still think a couple of these add a lot of great information, even though I don't really expect the user to ever touch this file. This is only used to execute the [execd shim](https://docs.influxdata.com/telegraf/v1.21/external_plugins/shim/), and should not be modified. Everything should be handled through the configs instead.

skejserjensen commented on cmd/main.go:
```diff
@@ -0,0 +1,59 @@
+	"time"
+
+	// TODO: import your plugins
+	_ "github.com/agneborn98/Arrow-Flight-ModelarDB-Output-Plugin/plugins/output/flight"
```
Do we want to order import by std, libraries, local like in ModelarDB? Also should this import statement not be updated to point to this repository?

agneborn98 replied:
So far I have followed the [style guide](https://github.com/influxdata/telegraf/blob/master/docs/developers/CODE_STYLE.md) in telegraf, which uses goimports to sort the imports.

Yes it should, but that is not possible until that file actually exists in the repository, which is when this PR has been merged.

skejserjensen replied:
I also think we should stick with the Go and Telegraf style guides and coding styles. I have left this comment unresolved as a reminder that the import statement should be updated in a final commit right before this PR is squashed and merged. Feel free to resolve this conversation when that commit have been pushed.

skejserjensen commented on cmd/main.go:
```diff
@@ -0,0 +1,59 @@
+	"github.com/influxdata/telegraf/plugins/common/shim"
+)
+
+var pollInterval = flag.Duration("poll_interval", 1*time.Second, "how often to send metrics")
```
Maybe add a comment documenting why these variables are global and not scoped to `main()` where they are used.

agneborn98 replied:
`main.go` is a template file from telegraf used to execute the `shim()`. 

There is no reasonable explanation for this, but is it really necessary for me to dig deeper into why they are outside the scope? 
My theory is that maybe the `flag` package wants it like that, but I think just leaving it as it is will be fine.

skejserjensen replied:
Well if they have to be in global scope then there is nothing we can do about it, however, a quick search indicates that it is [not a requirement](https://gobyexample.com/command-line-flags). I suggest we at least try to make them local and I quickly tried to move them into `main` and it seems to at least build without any errors.

agneborn98 replied:
The variables have been moved into the `main()` function and are working properly. Fixed.

skejserjensen commented on cmd/main.go:
```diff
@@ -0,0 +1,59 @@
+//
+// // now the shim.Run() call as below. Note the shim is only intended to run a single plugin.
+func main() {
+	// parse command line options
```
Maybe update all comments to start with an uppercase letter and end with a punctuation for consistency.

agneborn98 replied:
Fixed.

skejserjensen commented on go.mod:
```diff
@@ -0,0 +1,94 @@
+module github.com/agneborn98/Arrow-Flight-ModelarDB-Output-Plugin
```
Maybe regenerate this file so the correct module name is used.

agneborn98 replied:
See [this](https://github.com/ModelarData/Telegraf-Output-Apache-Arrow-Flight/pull/1#discussion_r947658389).

skejserjensen replied:
Like [here](https://github.com/ModelarData/Telegraf-Output-Apache-Arrow-Flight/pull/1#discussion_r950329186) I have left this conversation unresolved as a reminder that this file should be updated right before the PR is squashed and merged.

skejserjensen commented on cmd/main.go:
```diff
@@ -0,0 +1,59 @@
+package main
```
Descriptive [doc comments ](https://tip.golang.org/doc/comment)should probably be added to the packages, methods, types, etc.

agneborn98 replied:
Again, this is a template from telegraf. If the purpose here is to make a general-purpose Apache Arrow Flight plugin, I think we should leave these as is.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+// DO NOT REMOVE THE NEXT TWO LINES! This is required to embed the sampleConfig data.
+//
+//go:embed sample.conf
+var sampleConfig string
```
It is unclear why the contents of `sample.conf` is not just directly returned from `SampleConfig()` instead of being read from `sample.conf` at compile time when the users is never asked to changed it in the README. If the user should change both `telegraf.conf` and `sample.conf` it should probably be mentioned in the README.

agneborn98 replied:
The user is now asked to configure the `sample.conf` in the README.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+	return sampleConfig
+}
+
+type Flight struct {
```
As mentioned, maybe add some doc comments for the type and members to make their purpose more clear.

agneborn98 replied:
Added a doc comment explaining what the struct is. Explaining each individual member seems excessive, I haven't seen any other plugin do it this way.

skejserjensen replied:
If this is generally not done in the Go community it may be fine to wait and only add the remaining doc comments until `flight.go` becomes complex enough that it would be hard to comprehend without. What do you thing @CGodiksen and @chrthomsen?

CGodiksen replied:
For now I think it is fine to leave it as it is. I imagine this struct might become more complex when supporting different schemas so this can be reconsidered then.

agneborn98 replied:
I agree, let's close this one.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+	desc   *flight.FlightDescriptor
+}
+
+// Write should write immediately to the output, and not buffer writes
```
This comments sounds like part of the template, maybe rewrite it in the same style as those used in ModelarDB for consistency. For example document everywhere why a method may return an `error`.

agneborn98 replied:
Can you give an example of this? I have read through almost every other plugin written for telegraf and have yet to see anything like this.

skejserjensen replied:
Well if it is generally not done in the Go community it may be better to follow their coding style than what we do for Rust. What do you think @CGodiksen and @chrthomsen? For examples from ModelarDB, see, e.g., [`UncompressedDataManager.insert_message()`](https://github.com/ModelarData/ModelarDB-RS/blob/master/server/src/storage/uncompressed_data_manager.rs#L56) and [`ErrorBound::new()`](https://github.com/ModelarData/ModelarDB-RS/blob/master/server/src/models/mod.rs#L63).

CGodiksen replied:
I think the current comment seems fine. But again, if complexity of the code increases this can be reconsidered.

agneborn98 replied:
I agree, let's evaluate at a later time.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+		for _, field := range m.FieldList() {
+			s.values = append(s.values, float32(field.Value.(float64)))
+		}
+
```
Maybe remove the empty lines between multiple curly braces.

agneborn98 replied:
Fixed.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+	b := array.NewRecordBuilder(&s.pool, &s.schema)
+	defer b.Release()
+
+	b.Field(0).(*array.Int32Builder).AppendValues([]int32{1}, nil)
```
Maybe also document here that it is a known limitation of the plugin that only this schema is supported.

agneborn98 replied:
Added.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+	b := array.NewRecordBuilder(&s.pool, &s.schema)
+	defer b.Release()
+
+	b.Field(0).(*array.Int32Builder).AppendValues([]int32{1}, nil)
```
Maybe document why a tid is needed and why one is used.

agneborn98 replied:
I still don't quite grasp why tid is needed, I only did this implementation to adhere to the schema. Is it to support multivariate time-series? Can you quickly elaborate?

skejserjensen replied:
`tid` was the [unique auto-incrementing integer](https://github.com/ModelarData/ModelarDB/blob/master/src/main/java/dk/aau/modelardb/core/Partitioner.java#L52) that the legacy JVM-based version of ModelarDB used for uniquely identifying each time series. It is currently being removed from the Rust version so maybe add the comment "Unused and deprecated time series identifier." and then remove both the comment and code when it is no longer needed to match the schema in ModelarDB?

agneborn98 replied:
Sound good. Fixed.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+// Write should write immediately to the output, and not buffer writes
+// (Telegraf manages the buffer for you). Returning an error will fail this
+// batch of writes and the entire batch will be retried automatically.
+func (s *Flight) Write(metrics []telegraf.Metric) error {
```
Maybe use more descriptive variable names than just single letters through the plugin.

agneborn98 replied:
This is a method with a struct receiver and should always only be a single letter. 

I have yet to see it used any other way: 
[Example 1](https://gobyexample.com/methods)
[Example 2](https://github.com/influxdata/telegraf/blob/master/plugins/outputs/influxdb/influxdb.go)
[Example 3](https://www.golang-book.com/books/intro/9#section2)

agneborn98 replied:
I have changed the letter to add clarity.

skejserjensen replied:
Seem short [receiver](https://github.com/golang/go/wiki/CodeReviewComments#receiver-names) and [variable names](https://github.com/golang/go/wiki/CodeReviewComments#variable-names) are preferred when writing Go, so it is probably best to follow the current practice.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+// batch of writes and the entire batch will be retried automatically.
+func (s *Flight) Write(metrics []telegraf.Metric) error {
+
+	for _, m := range metrics {
```
Would it be possible to pre-allocate an `ArrayBuilder` for each column the length of `metrics` and than add each tid, timestamp, and value directly to these builders without temporarily adding them to arrays in `Flight` and thus reduce the number of allocations?

agneborn98 replied:
I have made some modifications to the function which I think satisfies what you explain here.

skejserjensen replied:
I think the changes improved the function, but maybe call [reserve()](https://pkg.go.dev/github.com/apache/arrow/go/arrow/array#RecordBuilder.Reserve) on the `RecordBuilder` with the size of `metrics` after adding the builders to ensure memory is only allocated once.

agneborn98 replied:
Fixed.

agneborn98 committed:
Retrieve flight descriptor from server

This commit modifies the plugin to retrieve the flight descriptor from the server instead of using the descriptor assigned in the config.

agneborn98 committed:
Update README.md

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+	err := s.writer.Write(rec)
+
+	if err != nil {
+		log.Fatal(err)
```
Maybe log the error and return it to Telegraf so it can retry transmitting the data points again?

agneborn98 replied:
Fixed. I have done this throughout the file so that when an error is encountered the error is returned to Telegraf.

agneborn98 committed:
Update README.md

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+	return nil
+}
+
+// Make any connection required here
```
Maybe expand this doc comment to document what the method does and add a punctuation at the end.

agneborn98 replied:
I don't understand this, the doc comment clearly states what the purpose of the method is. I have added punctuation.

skejserjensen replied:
I would have expected the comment to be something like "Connect to the Apache Arrow Flight server, retrieve the schema for the table to ingest metrics into, and initialize a DoPut stream." However, since the method is not that long and contains comments it is probably fine to leave the short comment as it matches the comment for `Close()`. Feel free to either directly resolve this conversation or update the comments for `Connect()` and `Close()` and then resolve the conversation.

agneborn98 replied:
Fixed.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+		log.Fatal(err)
+	}
+
+	s.ctx = context.Background()
```
Maybe document the properties of the context returned by [`Background()`](https://pkg.go.dev/context#Background).

agneborn98 replied:
What do you mean? Just add the explanation from [here](https://pkg.go.dev/context#Background)? Context is empty so it does not have any specific properties.

skejserjensen replied:
I think the added comment is fine.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+
+	deserializedSchema, err := flight.DeserializeSchema(retrievedSchema, memory.DefaultAllocator)
+
+	if err != nil {
```
Maybe return this `error` to Telegraf and document the result of returning an `error` like for `Write`.

agneborn98 replied:
Error is now being returned instead. The explanations in `Write` are not mine, but are from the template plugin in telegraf. I have added something similar in `Connect`.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+		log.Fatal(err)
+	}
+
+	s.pool = *memory.NewGoAllocator()
```
It is unclear why a new allocator is created here but `memory.DefaultAllocator` is used on Line 124 as it looks like `memory.DefaultAllocator` is a `GoAllocator` from the [documentation](https://pkg.go.dev/github.com/apache/arrow/go/arrow/memory#DefaultAllocator) and [code](https://github.com/apache/arrow/blob/master/go/arrow/memory/allocator.go#L33).

agneborn98 replied:
I don't understand what this means, I have to create an allocator to use in the following function, right? `NewGoAllocator()` and `memory.DefaultAllocator` are essentially the same thing, as the thing you linked to explains. 

I have switched it to `memory.DefaultAllocator` but an explanation as to why this is beneficial would be informative for me.

skejserjensen replied:
I simply did not understand why `memory.DefaultAllocator` could be used in one part of the code but a new allocator had to created using `NewGoAllocator()` on Line 138. I think the code is more clear now that it uses `memory.DefaultAllocator` everywhere. Although, maybe remove `Flight.pool` and use `memory.DefaultAllocator` directly?

agneborn98 replied:
Using the DefaultAllocator directly. Fixed.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+}
+
+func (s *Flight) Close() error {
+	// Close any connections here.
```
Maybe make these comments doc comments for the method.

agneborn98 replied:
Fixed.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,156 @@
+func (s *Flight) Close() error {
+	// Close any connections here.
+	// Write will not be called once Close is called, so there is no need to synchronize.
+	s.writer.Close()
```
Is it possible for `s.writer.Close()` to fail? If not, this should probably be documented, and if it can fail, the error should probably be returned.

agneborn98 replied:
Added error handling for `writer.Close()`.

skejserjensen commented on plugins/output/flight/flight_test.go:
```diff
@@ -0,0 +1,44 @@
+		t.Skip("Skipping integration testing in short mode")
+	}
+
+	container := testutil.Container{
```
Do we want to require Docker for running the integration tests? I think it is fine if the option for automatically creating and Apache Arrow Flight end point is available, but that it should be possible to execute the integration tests by manually creating an Apache Arrow Flight end point. What do you think @chrthomsen and @CGodiksen?

agneborn98 replied:
I could not get the docker integration to work, as I don't think any implementation of Apache Arrow Flight exists as an official image. I have resorted to a different test strategy.

This has been removed.

skejserjensen commented on plugins/output/flight/sample.conf:
```diff
@@ -0,0 +1,9 @@
+[[outputs.flight]]
+    ## Location to connect to
```
If '#' and '##' have the same effect maybe use '#' everywhere for consistency.

agneborn98 replied:
Fixed.

skejserjensen reviewed (changes requested):
While the general functionality is implemented, I think some cleanup is necessary before the code is merged. For example, I think doc comments should be added, that variable names longer than a single letter should be used, it should be checked that errors are handled whenever they are returned, that errors are properly returned to Telegraf everywhere instead of just terminating the process, and the comments that seems to be from the template should be removed or updated.

agneborn98 committed:
Reverted retrieving descriptor from server

agneborn98 committed:
README comment fixes

agneborn98 committed:
Comment fixes

agneborn98 committed:
Update flight.go

agneborn98 committed:
Fixes to comments

agneborn98 committed:
Update flight.go

agneborn98 committed:
Update main.go

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,63 @@
-# Telegraf-Output-Apache-Arrow-Flight
-External Telegraf plugin for writing time series to Apache Arrow Flight
+# Apache Arrow Flight External Output Plugin
```
Maybe write "... Telegraf Output Plugin".

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,63 @@
-External Telegraf plugin for writing time series to Apache Arrow Flight
+# Apache Arrow Flight External Output Plugin
+
+This plugin is a general purpose output plugin for the [Apache Arrow Flight protocol](https://arrow.apache.org/docs/format/Flight.html).
```
Maybe write "... Telegraf output plugin ..."

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,63 @@
+2. Build the binary:
+    * Windows: `go build -o binary/flight.exe cmd/main.go`
+    * Linux/macOS: `go build -o binary/flight cmd/main.go`
+3. Download the latest version of [telegraf](https://portal.influxdata.com/downloads/) for your platform. (To see which platform is needed, run the following command: `go env GOOS GOARCH`)
```
Replace all instances of "telegraf" with "Telegraf" since it is a name. Specify filenames should of course always have correct casing, e.g., "telegraf.conf".

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,63 @@
+    * Windows: `go build -o binary/flight.exe cmd/main.go`
+    * Linux/macOS: `go build -o binary/flight cmd/main.go`
+3. Download the latest version of [telegraf](https://portal.influxdata.com/downloads/) for your platform. (To see which platform is needed, run the following command: `go env GOOS GOARCH`)
+4. Extract the telegraf executable and configuration to the repository folder.
```
Maybe write "... Telegraf binaries ..." to match the language on the InfluxData website. This would also make it more clear what format to download Telegraf in, e.g., I assume the Docker Image is not supported when using these instructions.

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,63 @@
+4. Extract the telegraf executable and configuration to the repository folder.
+5. Configure the telegraf.conf file:
+   * In `telegraf.conf`, remove the comment in front of the tag `[[outputs.execd]]` and the option `command`.
+     * Windows: Assign the following to `command` so the resulting line becomes: `command = ["/path/to/flight.exe", "-config", "/path/to/sample.conf"]` 
```
Very minor thing, but maybe `command = ...` should be on a separate line.

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,63 @@
+    * Linux/macOS: `go build -o binary/flight cmd/main.go`
+3. Download the latest version of [telegraf](https://portal.influxdata.com/downloads/) for your platform. (To see which platform is needed, run the following command: `go env GOOS GOARCH`)
+4. Extract the telegraf executable and configuration to the repository folder.
+5. Configure the telegraf.conf file:
```
Maybe always format telegraf.conf as `telegraf.conf` so it is presented in a consistent manner.

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,63 @@
+
+## Configuration
+
+The following configuration is a [sample configuration](\plugins\output\flight\sample.conf) used to connect to the Arrow Flight Server and configure what table to insert the data to.
```
Very minor, but maybe change "... insert the data to" with "... insert the data into".

agneborn98 replied:
Fixed.

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,63 @@
+## Known issues and limitations
+
+* Currently, the plugin only implements support for the simplest schema supported by legacy JVM and current Rust versions of [ModelarDB](https://github.com/ModelarData/ModelarDB-RS), as listed above. Support for an arbitrary schema is planned.
+* `GetSchema()` is not compatible with the Rust implementation of Apache Arrow Flight, because a [bug](https://github.com/apache/arrow-rs/issues/2445) is present in the Rust implementation of Apache Arrow Flight where the schema is not serialized properly.
```
Maybe replace `GetSchema()` with "This plugin", as an end-user would need to know how Apache Arrow Flight and the plugin works to understand the implications of "`GetSchema()` is not compatible ..."

agneborn98 replied:
I agree, fixed.

skejserjensen replied:
Was this fix accidentally not included in the commit? [Known issues and limitations](https://github.com/ModelarData/Telegraf-Output-Apache-Arrow-Flight/tree/file-transfer#known-issues-and-limitations) still mention `GetSchema()` as far as I can see.

agneborn98 replied:
Good catch, forgot to save the README... Should be done now

skejserjensen commented on README.md:
```diff
@@ -1,2 +1,63 @@
+
+To run the tests: 
+1. Install the latest version of [Go](https://go.dev/doc/install).
+2. Start the Flight Server.
```
Maybe write "Apache Arrow Flight" server since we generally include Apache in the name of Apache projects. Also if this can be an arbitrary flight server, maybe write "Start **an** Apache Arrow Flight server" and include "on port ????" if it must run on a specific port.

agneborn98 replied:
I have fixed the naming. I also changed it to "Start an Apache Arrow Flight server", but the port just needs to match the configuration so I omitted that part.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,164 @@
+	builder := array.NewRecordBuilder(f.pool, &f.schema)
+	defer builder.Release()
+
+	//Currently, the plugin only implements support for the simplest schema
```
Missing space after // for multiple comments in the PR. Although I do not know if spaces are allow for "//go:embed sample.conf".

agneborn98 replied:
Fixed.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,164 @@
+
+	fmt.Println(rec)
+
+	err := f.writer.Write(rec)
```
Could Line 77 to Line 83 be rewritten to `return f.writer.Write(rec)`?

agneborn98 replied:
It absolutely is, fixed.

skejserjensen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,164 @@
+
+// Close any connections here.
+func (f *Flight) Close() error {
+	err := f.writer.Close()
```
Could this function be rewritten to `return f.writer.Close()`?

agneborn98 replied:
It is, has been fixed.

agneborn98 commented:
I have fixed every comment now, but am still waiting on response from @CGodiksen on some of the comments. When you have time, please make a review and resolve these comments if properly answered. Thank you.

agneborn98 committed:
Fixes for comments

agneborn98 committed:
Update README.md

skejserjensen reviewed (approved):
All of my comments have been resolved so I have approved the PR. The only very minor thing I noticed is the empty line between the comment and the code on Line 53 of [flight.go](https://github.com/ModelarData/Telegraf-Output-Apache-Arrow-Flight/blob/a7ba9302ad9c20ded5762f5f1e2de156df47e5c9/plugins/output/flight/flight.go#L53). However, it was unclear if this empty line was intentional and even if it isn't, I do not think it is worth going through an additional round of code review to discuss how to handle this line. The remaining unresolved conversations require input from @CGodiksen and/or @chrthomsen before they can be resolved. @CGodiksen and @chrthomsen, feel free to resolve the remaining conversations if you agree with the current state of the code or after it has been updated.

CGodiksen commented on README.md:
```diff
@@ -1,2 +1,65 @@
+
+## Arrow Schema
+
+The plugin currently supports inserting data points with the following schema:
```
maybe "...outputting data points..." instead.

agneborn98 replied:
Fixed.

CGodiksen commented on README.md:
````diff
@@ -1,2 +1,65 @@
+```go
+schema:
+  fields: 3
+    - tid: type=int32 //Unused and deprecated time series identifier.
````
Missing a space at the start of the comment.

agneborn98 replied:
Fixed.

CGodiksen commented on cmd/main.go:
```diff
@@ -0,0 +1,39 @@
+
+func main() {
+
+	var pollInterval = flag.Duration("poll_interval", 1*time.Second, "how often to send metrics")
```
Maybe change to "1 * time.Second"

agneborn98 replied:
I also wanted this, but the go formatter keeps reverting it every time I save the file...

CGodiksen commented on cmd/main.go:
```diff
@@ -0,0 +1,39 @@
+
+func main() {
+
+	var pollInterval = flag.Duration("poll_interval", 1*time.Second, "how often to send metrics")
```
Maybe mention in the help message what the expected time unit is.

agneborn98 replied:
I'm going to need some elaboration. 

Should I just write what units are possible to enter as arguments? (Anything can be entered here) 

Or do you mean what the default duration is?

CGodiksen replied:
I am not sure how this works. Do you not just enter a number? If just entering a number it can be hard to know if entering "1" means 1 second or 1 millisecond for example. But if you enter an actual duration then this might not be an issue.

agneborn98 replied:
Ah, I see. I have made a quick change that I think satisfies what you suggested.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,149 @@
+	"github.com/influxdata/telegraf/plugins/outputs"
+)
+
+//go:embed sample.conf
```
Missing a space at the beginning of this comment.

agneborn98 replied:
As explained [here](https://pkg.go.dev/cmd/compile#hdr-Compiler_Directives), this is not a literal comment but actually, a compiler directive used to embed that specific file. Their missing space at the beginning of the comment is recognized explicitly as directives.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,149 @@
+	// supported by legacy JVM and current Rust versions of [ModelarDB](https://github.com/ModelarData/ModelarDB-RS),
+	// as listed below. Support for an arbitrary schema is planned.
+
+	tidBuilder := builder.Field(0).(*array.Int32Builder) //Unused and deprecated time series identifier.
```
Missing a space at the beginning of this comment.

agneborn98 replied:
Fixed.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,149 @@
+		timeBuilder.AppendValues([]arrow.Timestamp{arrow.Timestamp(timeInt)}, nil)
+
+		for _, field := range m.FieldList() {
+			valueBuilder.AppendValues([]float32{float32(field.Value.(float64))}, nil)
```
If all fields are appended to the same value builder, does this not break when using more than 1 field? I would assume using two fields would mean that the value builder will be 2x the length of the other builders.  If this code only works when using a single field, maybe remove it and just assume that there only is one field.

agneborn98 replied:
True, but as we have discussed above and as the README says, this is currently very specific to our use. The plugin will eventually have to be able to support creating multiple fields in the more generalized update to this plugin.

CGodiksen replied:
I still think it should be removed and replaced by a single `valueBuilder.AppendValues` call with the first field for now. The current code makes it look like it supports multiple fields, but will break if trying to use multiple fields.

agneborn98 replied:
Fixed.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,149 @@
+// it will be written to the output and the plugin will attempt to restart.
+func (f *Flight) Connect() error {
+
+	// Create a new connection to the grpc using the given target.
```
Maybe "Create a new connection to the gRPC server using..." instead.

agneborn98 replied:
Fixed.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,149 @@
+	// Initialize a new Flight Service Client using the client API and add it to the Flight struct.
+	f.client = flight.NewFlightServiceClient(conn)
+
+	if err != nil {
```
It does not seem like `err` is changed since the last `if err != nil`. Maybe remove this.

agneborn98 replied:
Fixed.

CGodiksen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,149 @@
+
+	f.schema = *deserializedSchema
+
+	if err != nil {
```
It does not seem like `err` is changed since the last `if err != nil`. Maybe remove this.

agneborn98 replied:
Fixed.

agneborn98 committed:
Minor fixes

CGodiksen reviewed (commented):
Most of the fixes were approved. There are just some minor things left.

agneborn98 committed:
Fix FieldList

agneborn98 committed:
Update main.go

agneborn98 committed:
Final fixes

chrthomsen commented on README.md:
````diff
@@ -1,2 +1,65 @@
+```go
+schema:
+  fields: 3
+    - tid: type=int32 // Unused and deprecated time series identifier.
````
Where is it deprecated? And if it is "deprecated", its use is discouraged, but here we include it... So why include it and why say that it is a time series identifier if it anyways is unused and deprecated?

agneborn98 replied:
See [this](https://github.com/ModelarData/Telegraf-Output-Apache-Arrow-Flight/pull/1#discussion_r950370228) comment.

chrthomsen replied:
OK, let's leave it for now and then remove it when possible

chrthomsen commented on README.md:
```diff
@@ -1,2 +1,65 @@
+      `command = ["/path/to/flight.exe", "-config", "/path/to/sample.conf"]` 
+     * Linux/macOS: Assign the following to `command` so the resulting line becomes:  
+      `command = ["/path/to/flight", "-config", "/path/to/sample.conf"]` 
+   * Configure any input plugin to consume metrics. (the metric must adhere to the schema presented in this README)
```
Remove full stop or capital T

agneborn98 replied:
Fixed.

chrthomsen commented on README.md:
```diff
@@ -1,2 +1,65 @@
+7. Run the plugin using Telegraf: `telegraf --config telegraf.conf --input-filter chosen_input_plugin --output-filter execd`
+
+To run the tests: 
+1. Install the latest version of [Go](https://go.dev/doc/install).
```
Is it possible to run the tests without doing the steps above (where Go also was installed)? If it is, then this is fine. If not, there is no need to Install the latest version of Go again here

agneborn98 replied:
Yeah, exactly. This is possible and is why I included repeating that step in the testing.

chrthomsen commented on README.md:
````diff
@@ -1,2 +1,65 @@
+The following configuration is a [sample configuration](\plugins\output\flight\sample.conf) used to connect to the Arrow Flight Server and configure what table to insert the data into.
+
+```toml @sample.conf
+## Configuration for Arrow Flight to send metrics to.
````
for "Arrow Flight" or "Arrow Flight Server"?

agneborn98 replied:
Arrow Flight Client, and some minor modifications. Is this okay?

chrthomsen commented on README.md:
```diff
@@ -1,2 +1,65 @@
+    
+    ## Name of the table to store the metrics in.
+    ## example: table = "data"
+    table = ""
```
why is table = "" when this is a sample config? Is empty string allowed here - and if yes, what does it mean?

agneborn98 replied:
You are right, the sample config should include a sample on this one. Fixed.

chrthomsen commented on cmd/main.go:
```diff
@@ -0,0 +1,39 @@
+
+func main() {
+
+	var pollInterval = flag.Duration("poll_interval", 1*time.Second, "how often to send metrics(in seconds)")
```
space before "(in seconds)"

agneborn98 replied:
Fixed.

chrthomsen commented on cmd/main.go:
```diff
@@ -0,0 +1,39 @@
+func main() {
+
+	var pollInterval = flag.Duration("poll_interval", 1*time.Second, "how often to send metrics(in seconds)")
+	var pollIntervalDisabled = flag.Bool("poll_interval_disabled", false, "set to true to disable polling. You want to use this when you are sending metrics on your own schedule")
```
"You want to use this when" -> "Use this when" ?

agneborn98 replied:
Fixed.

chrthomsen commented on cmd/main.go:
```diff
@@ -0,0 +1,39 @@
+
+	err = shimLayer.LoadConfig(configFile)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Err loading input: %s\n", err)
```
Why "Err" instead of "Error"? Is that a convention in Telegraf or Go or ...?

agneborn98 replied:
Exactly, it is a [convention](https://go.dev/blog/error-handling-and-go) used throughout golang.

chrthomsen replied:
But why should the user see the error message "Err ..." instead of "Error ..."? (See the first 'word' in the second argument to Fprintf)

agneborn98 replied:
Ah, now I understand. Fixed.

chrthomsen commented on plugins/output/flight/flight.go:
```diff
@@ -0,0 +1,141 @@
+var sampleConfig string
+
+// SampleConfig returns the configuration of the Apache Arrow Flight plugin.
+func (*Flight) SampleConfig() string {
```
I'm a little confused here. What is this for? Why is it a *sample* rather than an actual configuration? Why do we ask the user to make configurations in a file called "sample"?

agneborn98 replied:
I can see why it would be confusing, but this is the [convention](https://github.com/influxdata/telegraf/blob/master/docs/CONFIGURATION.md#plugins) used in Telegraf. I believe it is because this is attached to the main telegraf.conf file once the plugin has been committed to the main Telegraf repository. It think it doesn't really matter what it is called because I clearly state in the README that this should be configured. If the user doesn't want to do that, the user can technically copy these lines into any config, modify the `command = ["/path/to/flight.exe", "-config", "/path/to/sample.conf"]` to point to that and use that instead.

chrthomsen replied:
OK, I think it is a bit weird to have the actual configuration in a sample configuration. But if that is how it typically is done, then let's continue like that...

chrthomsen commented on cmd/main.go:
```diff
@@ -0,0 +1,39 @@
+	}
+
+	if err := shimLayer.Run(*pollInterval); err != nil {
+		fmt.Fprintf(os.Stderr, "Err: %s\n", err)
```
Same question as above

agneborn98 replied:
Exactly, it is a [convention](https://go.dev/blog/error-handling-and-go) used throughout golang.

chrthomsen replied:
And same comment again - why should the user see "Err:" instead of "Error:"?

agneborn98 replied:
Ah, now I understand. Fixed.

agneborn98 committed:
Comment fixes

agneborn98 committed:
Update main.go
